### ISSUE TITLE
secretgen-controller installed by default in TCE unmanaged-clusters

### ISSUE THREAD

**Issue by kartiklunkad26** [vmware-tanzu/community-edition#3817]
## Feature Request

To set-up a developer environment with App Toolkit, we require the user to create a placeholder secret with the `tanzu secret` plugin that then is propagated to all namespaces. 

We require `secretgen-controller` to help us enable this outcome. 

## Describe alternatives you've considered

The alternative would be to install secretgen-controller manually by the user which is not a great experience OR App Toolkit installs it which I don't think is appropriate given secretgen-controller is a common utility for additional packages.

**Comment by joshrosso** [comment]
@kartiklunkad26 so the expectation is that `secretgen-controller` is installed by default in all unmanged-clusters?

**Comment by jorgemoralespou** [comment]
@joshrosso secretgen is available on managed clusters by default, so I would consider that expectations from what "tanzu" provides in the clusters by default should be aligned. As `uc` does provide what's been called `tanzufying` clusters "capability" so in that regard we should define what `tanzufying` means (which I consider having feature parity on user-space stuff on the cluster)

**Comment by kartiklunkad26** [comment]
secretgen-controller is closer to kapp-controller in the sense it's generic functionality and not related to the "App-Toolkit" package. I consider it as the minimum capabilities in a cluster, and hence the ask to have it be available on all `unmanaged-clusters`. 

My option B would be to not leverage secretgen-controller at all rather than including it in App Toolkit or asking users to install it separately. 

I can check with my team the impact of not using `secretgen-controller` and reply back on this thread.

**Comment by kartiklunkad26** [comment]
Not using `secretgen-controller` means the users will have to apply a bunch of YAML to be able to use App Toolkit :(

**Comment by jorgemoralespou** [comment]
@joshrosso If the issue is `A core issue with secretgen controller is that it does not ship an arm64 container image` We need to push Carvel team to do so, as we do for kapp-controller. 

https://github.com/vmware-tanzu/carvel-secretgen-controller/issues/57

> A definition of what qualifies a cluster as “tanzu” is definitely needed.

https://github.com/vmware-tanzu/community-edition/issues/3845

I recommend c.

c. Get the carvel team to produce arm64 images so that we can `tanzuify` arm64 based clusters.

**Comment by kartiklunkad26** [comment]
> In summary, it is either break arm support or take a UX hit for a subset of users.

The impact of breaking arm support is impacting users who use arm support (a subset of TCE users) which as @jorgemoralespou is saying is pretty tiny especially until ARM support is a full fledged feature in TCE.

**Comment by joshrosso** [comment]
@jorgemoralespou @kartiklunkad26 appreciate your follow-up and input.

Unfortunately, I don't quite see this as ya'll do.

* I'm unwilling to send broken workloads (by default) into clusters based on architectures we support.
* If we feel the arm user base is so small that we're willing to break them (for every cluster bootstrap they do) then we should remove arm support altogether.

Extra data points:

* My opinions on the above would be different if AppToolkit was the primary use case for unmanaged-cluster.
* VMware controlled projects should simply produce arm-based images, as most upstream projects do.

Apologies that we're at a bit of an impasse (disagreement) here. Can ya'll please get with @vkorbes to voice your concerns here? Thanks all!

**Comment by kartiklunkad26** [comment]
Fair on having different opinions on the topic. I think my (and our) perspective is shared here. 

I would ask for you and @vkorbes to let me know the ultimate decision, so I can accordingly proceed!

**Comment by kartiklunkad26** [comment]
@seemiller @joshrosso (and anyone else) - How would the alternative work? 

- Is secretgen-controller available as a package in TCE Package Repository today? For us to install it as part of App Toolkit.

**Comment by jpmcb** [comment]
@kartiklunkad26 - it doesn't look like secret gen controller is in our package repo today.

```
❯ k get packages -A | grep secret
❯
```

But if we added it, as part of the app-toolkit metapackage, it could install and reconcile the secret-gen-controller as part of the all the other packages it installs.

**Comment by kartiklunkad26** [comment]
@jpmcb I don't think I follow. I think App Toolkit installs all packages that are available in the TCE Package Repo. So if it isn't present in TCE Package Repo, we wouldn't be able to install it. 

cc: @krisapplegate Kris - can you help please?

**Comment by krisapplegate** [comment]
I think @jpmcb is saying that we would need to get it available as a package and then include it in app-toolkit. The problem is that it's not there as a package in 0.11.0

**Comment by jpmcb** [comment]
> I think App Toolkit installs all packages that are available in the TCE Package Repo

That doesn't appear to be the case.

looks like app-toolkit is installing cartographer, cert-manager, contour, fluxcd, knative, and kpack.

```
❯ k get packageinstalls -A
NAMESPACE                   NAME                       PACKAGE NAME                                          PACKAGE VERSION   DESCRIPTION           AGE
tanzu-package-repo-global   cartographer               cartographer.community.tanzu.vmware.com               0.2.2             Reconcile succeeded   23m
tanzu-package-repo-global   cert-manager               cert-manager.community.tanzu.vmware.com               1.6.1             Reconcile succeeded   24m
tanzu-package-repo-global   contour                    contour.community.tanzu.vmware.com                    1.19.1            Reconcile succeeded   24m
tanzu-package-repo-global   fluxcd-source-controller   fluxcd-source-controller.community.tanzu.vmware.com   0.21.2            Reconcile succeeded   24m
tanzu-package-repo-global   knative-serving            knative-serving.community.tanzu.vmware.com            1.0.0             Reconcile succeeded   24m
tanzu-package-repo-global   kpack                      kpack.community.tanzu.vmware.com                      0.5.1             Reconcile succeeded   24m
tkg-system                  app-toolkit                app-toolkit.community.tanzu.vmware.com                0.1.0             Reconcile succeeded   25m
tkg-system                  cni                        calico.community.tanzu.vmware.com                     3.22.1            Reconcile succeeded   25m
tkg-system                  external-dns               external-dns.community.tanzu.vmware.com               0.10.0            Reconcile succeeded   25m
```

Which is a small subset of all available packages:

```
❯ k get packages -A
NAMESPACE                   NAME                                                         PACKAGEMETADATA NAME                                  VERSION    AGE
tanzu-package-repo-global   app-toolkit.community.tanzu.vmware.com.0.1.0                 app-toolkit.community.tanzu.vmware.com                0.1.0      16m26s
tanzu-package-repo-global   cartographer.community.tanzu.vmware.com.0.2.2                cartographer.community.tanzu.vmware.com               0.2.2      16m26s
tanzu-package-repo-global   cert-injection-webhook.community.tanzu.vmware.com.0.1.0      cert-injection-webhook.community.tanzu.vmware.com     0.1.0      16m26s
tanzu-package-repo-global   cert-manager.community.tanzu.vmware.com.1.3.3                cert-manager.community.tanzu.vmware.com               1.3.3      16m25s
tanzu-package-repo-global   cert-manager.community.tanzu.vmware.com.1.4.4                cert-manager.community.tanzu.vmware.com               1.4.4      16m25s
tanzu-package-repo-global   cert-manager.community.tanzu.vmware.com.1.5.4                cert-manager.community.tanzu.vmware.com               1.5.4      16m25s
tanzu-package-repo-global   cert-manager.community.tanzu.vmware.com.1.6.1                cert-manager.community.tanzu.vmware.com               1.6.1      16m25s
tanzu-package-repo-global   contour.community.tanzu.vmware.com.1.18.1                    contour.community.tanzu.vmware.com                    1.18.1     16m25s
tanzu-package-repo-global   contour.community.tanzu.vmware.com.1.19.1                    contour.community.tanzu.vmware.com                    1.19.1     16m25s
tanzu-package-repo-global   contour.community.tanzu.vmware.com.1.20.1                    contour.community.tanzu.vmware.com                    1.20.1     16m26s
tanzu-package-repo-global   external-dns.community.tanzu.vmware.com.0.10.0               external-dns.community.tanzu.vmware.com               0.10.0     16m24s
tanzu-package-repo-global   external-dns.community.tanzu.vmware.com.0.8.0                external-dns.community.tanzu.vmware.com               0.8.0      16m24s
tanzu-package-repo-global   fluent-bit.community.tanzu.vmware.com.1.7.5                  fluent-bit.community.tanzu.vmware.com                 1.7.5      16m24s
tanzu-package-repo-global   fluxcd-source-controller.community.tanzu.vmware.com.0.21.2   fluxcd-source-controller.community.tanzu.vmware.com   0.21.2     16m24s
tanzu-package-repo-global   gatekeeper.community.tanzu.vmware.com.1.0.0                  gatekeeper.community.tanzu.vmware.com                 1.0.0      16m24s
tanzu-package-repo-global   gatekeeper.community.tanzu.vmware.com.3.7.0                  gatekeeper.community.tanzu.vmware.com                 3.7.0      16m24s
tanzu-package-repo-global   grafana.community.tanzu.vmware.com.7.5.11                    grafana.community.tanzu.vmware.com                    7.5.11     16m26s
tanzu-package-repo-global   grafana.community.tanzu.vmware.com.7.5.7-1                   grafana.community.tanzu.vmware.com                    7.5.7-1    16m26s
tanzu-package-repo-global   harbor.community.tanzu.vmware.com.2.2.3                      harbor.community.tanzu.vmware.com                     2.2.3      16m26s
tanzu-package-repo-global   harbor.community.tanzu.vmware.com.2.3.3                      harbor.community.tanzu.vmware.com                     2.3.3      16m26s
tanzu-package-repo-global   knative-serving.community.tanzu.vmware.com.0.22.0            knative-serving.community.tanzu.vmware.com            0.22.0     16m26s
tanzu-package-repo-global   knative-serving.community.tanzu.vmware.com.0.26.0            knative-serving.community.tanzu.vmware.com            0.26.0     16m26s
tanzu-package-repo-global   knative-serving.community.tanzu.vmware.com.1.0.0             knative-serving.community.tanzu.vmware.com            1.0.0      16m26s
tanzu-package-repo-global   kpack.community.tanzu.vmware.com.0.5.0                       kpack.community.tanzu.vmware.com                      0.5.0      16m26s
tanzu-package-repo-global   kpack.community.tanzu.vmware.com.0.5.1                       kpack.community.tanzu.vmware.com                      0.5.1      16m26s
tanzu-package-repo-global   local-path-storage.community.tanzu.vmware.com.0.0.19         local-path-storage.community.tanzu.vmware.com         0.0.19     16m25s
tanzu-package-repo-global   local-path-storage.community.tanzu.vmware.com.0.0.20         local-path-storage.community.tanzu.vmware.com         0.0.20     16m25s
tanzu-package-repo-global   multus-cni.community.tanzu.vmware.com.3.7.1                  multus-cni.community.tanzu.vmware.com                 3.7.1      16m25s
tanzu-package-repo-global   prometheus.community.tanzu.vmware.com.2.27.0-1               prometheus.community.tanzu.vmware.com                 2.27.0-1   16m25s
tanzu-package-repo-global   velero.community.tanzu.vmware.com.1.6.3                      velero.community.tanzu.vmware.com                     1.6.3      16m25s
tanzu-package-repo-global   velero.community.tanzu.vmware.com.1.7.1                      velero.community.tanzu.vmware.com                     1.7.1      16m25s
tanzu-package-repo-global   velero.community.tanzu.vmware.com.1.8.0                      velero.community.tanzu.vmware.com                     1.8.0      16m24s
tanzu-package-repo-global   whereabouts.community.tanzu.vmware.com.0.5.0                 whereabouts.community.tanzu.vmware.com                0.5.0      16m26s
tkg-system                  antrea.community.tanzu.vmware.com.0.13.3                     antrea.community.tanzu.vmware.com                     0.13.3     16m36s
tkg-system                  calico.community.tanzu.vmware.com.3.22.1                     calico.community.tanzu.vmware.com                     3.22.1     16m36s
tkg-system                  local-path-storage.community.tanzu.vmware.com.0.0.20         local-path-storage.community.tanzu.vmware.com         0.0.20     16m36s
tkg-system                  metrics-server.community.tanzu.vmware.com.0.5.1              metrics-server.community.tanzu.vmware.com             0.5.1      16m36s
tkg-system                  pinniped.community.tanzu.vmware.com.0.12.1                   pinniped.community.tanzu.vmware.com                   0.12.1     16m36s
```

If secret-gen-controller was available in the package repository, app-toolkit could add it as another `PackageInstall` and it would automatically reconcile along side the other packages.

**Comment by joshrosso** [comment]
The source of confusion here may be that: 

`secretgen-controller` is part of the `core` package repository, not the TCE package (user-managed) repository.

It's not included in the core package repo in unmanaged-clusters because we trim down the core package repository.

I'm working through an alternative option involving forking secretgen-controller for our next release.

Please check back tomorrow.

**Comment by kartiklunkad26** [comment]
Okay. @joshrosso Let's talk through all the options once you have it articulated. 

We may decide to drop the feature altogether, if the collective effort is _too much_, and live with the user impact. 

The impact of _doing nothing_ would be a user using kubectl to configure a namespace for usage with App Toolkit.

**Comment by joshrosso** [comment]
Here are the options we have.

1. **Use a fork of `secretgen-controller` and have it installed by default**
    * Fork supports amd64 and arm64
    * We can move back to upstream once https://github.com/vmware-tanzu/carvel-secretgen-controller/issues/57 is resolved.
    * This introduces net-new technical debt
    * See PR with demo: https://github.com/vmware-tanzu/community-edition/pull/3901
2. **Add existing `secretgen-controller` to uc package repo; don't install by default**
    * No fork required
    * Users of things like App Toolkit will need to `tanzu package install` (or equivalent) secretgen-controller
    * Gives users the option to install, and ARM users can choose not to, rather than have a failing workload automatically deployed.
3. **Add existing `secretgen-controller` to uc package repo; install by default**
    * No fork required
    * ARM users will have failing workload after bootstrap
4.  **Do nothing until upstream provides arm images and our tanzu packages are updated**
    * Same UX as today

I highly recommend `1` or `2`. I strongly recommend against `3`.

@kartiklunkad26 @vkorbes to talk about the final call.

**Comment by kartiklunkad26** [comment]
I’m good with option 2! @vkorbes let us know what you think. 

If that’s the final path, let us know what timeline that would be in, as we would then look to include it in App Toolkit for v0.12.

**Comment by kartiklunkad26** [comment]
And thank you @joshrosso for your fast and clear and crisp summary of the possible options

**Comment by jorgemoralespou** [comment]
I know I haven't been asked for an opinion and I'm not a stakeholder but I recommend option 1 for consistency of what a Tanzu cluster should look like and have. I understand the technical debt problem but since there's already the same problem with kapp-controller, I would consider this to be the best option.
Option 2 creates difference in what users expect from tanzu clusters (managed vs unmanaged) and I'm documentation as it'll need to be written with an interim solution that once arm64 is available should be removed (if the goal of consistent tanzu clusters still exist).
I understand @kartiklunkad26 preferring option 2 mostly because he probably is looking at AppToolkit experience alone, but we should not.

**Comment by joshrosso** [comment]
I do agree with @jorgemoralespou's sentiment that we're already doing this for `kapp-controller`, so it may be sensible to do it for secretgen in the near-term :+1: .

**Comment by vkorbes** [comment]
Correct me if I'm wrong, but this is what I'm getting from this thread:

- `secretgen-controller` is a common utility used by packages...
- ...and those packages don't work on ARM.
- Building `secretgen-controller` for ARM won't enable us to use those packages.
- We don't want extra debt for little tangible benefit.

- Having to install `secretgen-controller` manually is not great...
- ...but we don't want broken setups installed by default either.
- Counterpoint: `--profile` is coming.

My take:

- We go with option 2: "Add existing `secretgen-controller` to uc package repo; don't install by default."
- We provide a `--profile` that includes `secretgen-controller`, as a UX improvement.
- We talk to the Carvel team about ARM builds for `.next`.

Am I missing anything?

**Comment by vkorbes** [comment]
(Will edit this comment with clarifications in a bit.)

**Comment by joshrosso** [comment]
This work is now targeting [v0.13.0](https://github.com/vmware-tanzu/community-edition/milestone/15).

For `v0.12.0`, we will now target completion the following **instead**:

* https://github.com/vmware-tanzu/community-edition/issues/3909

**Comment by jorgemoralespou** [comment]
@joshrosso so in 0.13 the secretgen controller will move from one packagerepository (TCE) to another one (core)? Will then live in both repos? Does that work on Carvel? What happens to users that will install it manually but then have it installed by default by the core package? Will the apptoolkit installation fails as it's trying to install an already installed package? Will the apptoolkit package need to ship a new version and mandate users to move to the new version if the cluster is upgraded?
Will AppToolkit with the secretgen provided work on managed clusters in 0.12?

**Comment by joshrosso** [comment]
> so in 0.13 the secretgen controller will move from one packagerepository (TCE) to another one (core)?

We will ship secretgen in `uc`'s core repo for both `v0.12.0` and `v0.13.0`.

![image](https://user-images.githubusercontent.com/6200057/161796887-352cdef9-634f-4a20-a567-f5f5bbb50569.png)

> Will then live in both repos?

It will only live in `core`.

> What happens to users that will install it manually but then have it installed by default by the core package?

It will only live in the `core` repo. This is the same problem we have today. For example, we install `calico` by default but nothing is preventing a user from trying to do a package install of `antrea`.

> Will the apptoolkit installation fails as it's trying to install an already installed package?

That's up to apptoolkit. If you're concerned with this UX, please get with @kartiklunkad26 to discuss it. Thank you!

> Will the apptoolkit package need to ship a new version and mandate users to move to the new version if the cluster is upgraded?

Unmanaged clusters does not and will not support upgrades. If you're concerned with this UX, please get with @kartiklunkad26 to discuss it. Thank you!

> Will AppToolkit with the secretgen provided work on managed clusters in 0.12?

We will continue to ship clusters consistent with TKG 1.5.X. Assuming app toolkit is able to work with that version (it is `v0.7.1`) then yes. If you're concerned with this UX, please get with @kartiklunkad26 to discuss it. Thank you!

**Comment by jorgemoralespou** [comment]
@joshrosso sorry, I misread the repository and I took it was going to be in the TCE user's provided repo. Now I see that there's no problem from TCE side, only from App Toolkit.

Thanks for clarification

**Comment by kartiklunkad26** [comment]
> Will the apptoolkit installation fails as it's trying to install an already installed package? 
Will include this as a requirement (if feasible). @krisapplegate can we include detection logic before we install `secretgen-controller`? 

> Will the apptoolkit package need to ship a new version and mandate users to move to the new version if the cluster is upgraded?
If we are to build the detection logic above, then we won't need to "mandate" users to move to newer version per-se. We will look to remove it from our package and ship a new App Toolkit version, once secretgen-controller is available by default. 


> Will AppToolkit with the secretgen provided work on managed clusters in 0.12?
Again, if we do the first question right, then this should _just_ work seamlessly.

**Comment by jorgemoralespou** [comment]
@kartiklunkad26 curious how this could be feasible given this is a long time requested feature in kapp-controller that it's not implemented: https://github.com/vmware-tanzu/carvel-kapp-controller/issues/410

**Comment by kartiklunkad26** [comment]
@jorgemoralespou I didn't know that. Alternative would be to ask users to use the `excluded_packages` feature in App Toolkit to exclude secretgen-controller that might be already installed in their cluster.

**Comment by jorgemoralespou** [comment]
So, the experience you propose @kartiklunkad26 is: 
try to install, if it falls, exclude the package in the config? Are you going to have all these steps in the docs? 🤣

**Comment by kartiklunkad26** [comment]
**TLDR:** 

```
👍 for keeping the current plan of having secretgen-controller in UC package repo for TCE v0.12, but prefer it installed by default in future TCE releases. 
```
---
**UX for TCE v0.12**
1. `tanzu package install secretgen-controller`
2. `tanzu secret add abc`
3. `tanzu package install app-toolkit`

---
**Learnings**

@krisapplegate spent some time thinking further about our approach, and learned that we cannot include `secretgen-controller` as part of App Toolkit. 

Below are the two steps that need to happen for a developer to create workloads. 

1. User creates a secret `abc` and exports to all namespaces using `tanzu secret` plugin and `secretgen-controller` package
4. App Toolkit creates the placeholder objects that references `abc` secret ^. 

As you can see App Toolkit needs to reference to a secret that already needs to exist, and that secret needs secretgen-controller to be installed to be created.

**Comment by cmoulliard** [comment]
> * more?

certificate manager
private registry (TLS Harbor with Contour route)

**Comment by cmoulliard** [comment]
> I recommend c.
> 
> c. Get the carvel team to produce arm64 images so that we can `tanzuify` arm64 based clusters

I fully agree as seret is a key core feature, needed by several packages like this TCE command which is available OOTB `tanzu secret registry add`

**Comment by ynamadi** [comment]
Hey, I just did these two commands on my M1 and it seem to install the secretgen-controller and add registry-credentials. I am on an unmanaged cluster using local docker. 

```
tanzu package install secretgen-controller --package-name secretgen-controller.community.tanzu.vmware.com --version 0.7.1 --namespace tkg-system
```

```
tanzu secret registry add registry-credentials --server https://index.docker.io/v1/ --username <USERNAME> --password <PASSWORD> --export-to-all-namespaces
```
